interval/search_tree_test.go

```diff
@@ -173,6 +178,496 @@ func TestMultiValueSearchTree_IsEmpty(t *testing.T) {
+				st.Insert(17, 19, "node1")
+				st.Insert(5, 8, "node2")
+				st.Insert(21, 24, "node3")
+				st.Insert(21, 24, "node4")
+				st.Insert(4, 4, "node5")
+
+				return st
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			mustTestMultiValueSearchTree_DecodingError(t, tt.tree())
+		})
+	}
+}
+
+func mustTestMultiValueSearchTree_DecodingError(t *testing.T, st1 *MultiValueSearchTree[string, int]) {
+	t.Helper()
+
+	st2 := NewMultiValueSearchTree[string, string](func(x, y string) int { return strings.Compare(x, y) })
+
+	b := mustEncodeMultiValueTree(t, st1)
+	r := bufio.NewReader(&b)
+	dec := gob.NewDecoder(r)
+
+	err := dec.Decode(&st2)
+	if err == nil {
+		t.Fatal("got unexpected <nil> error; want not nil")
+	}
+}
+
+func TestMultiValueSearchTree_DecodingTypeMismatchError(t *testing.T) {
+	tests := []struct {
+		name string
+		tree func() *MultiValueSearchTree[string, int]
+	}{
+		{
+			name: "with default options",
+			tree: func() *MultiValueSearchTree[string, int] {
+				st := NewMultiValueSearchTree[string, int](func(x, y int) int { return x - y })
+				st.Insert(17, 19, "node1")
+				st.Insert(5, 8, "node2")
+				st.Insert(21, 24, "node3")
+				st.Insert(21, 24, "node4")
+				st.Insert(4, 4, "node5")
+
+				return st
+			},
+		},
+		{
+			name: "with default options & empty",
+			tree: func() *MultiValueSearchTree[string, int] {
+				return NewMultiValueSearchTree[string, int](func(x, y int) int { return x - y })
+			},
+		},
+		{
+			name: "with interval point",
+			tree: func() *MultiValueSearchTree[string, int] {
+				st := NewMultiValueSearchTreeWithOptions[string, int](func(x, y int) int { return x - y }, TreeWithIntervalPoint())
+				st.Insert(17, 19, "node1")
+				st.Insert(5, 8, "node2")
+				st.Insert(21, 24, "node3")
+				st.Insert(21, 24, "node4")
+				st.Insert(4, 4, "node5")
+
+				return st
+			},
+		},
+		{
+			name: "with interval point & empty",
+			tree: func() *MultiValueSearchTree[string, int] {
+				return NewMultiValueSearchTreeWithOptions[string, int](func(x, y int) int { return x - y }, TreeWithIntervalPoint())
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			mustTestMultiValueSearchTree_DecodingTypeMismatchError(t, tt.tree())
+		})
+	}
+}
+
+func mustTestMultiValueSearchTree_DecodingTypeMismatchError(t *testing.T, st1 *MultiValueSearchTree[string, int]) {
+	t.Helper()
+
+	st2 := NewMultiValueSearchTree[string, string](func(x, y string) int { return strings.Compare(x, y) })
+
+	b := mustEncodeMultiValueTree(t, st1)
+	r := bufio.NewReader(&b)
+	dec := gob.NewDecoder(r)
+
+	err := dec.Decode(&st2)
+	if err == nil {
+		t.Fatal("got unexpected <nil> error; want not nil")
+	}
+
+	if err.Error() != `interval: cannot decode type "MultiValueSearchTree[string, int]" into type "MultiValueSearchTree[string, string]"` {
+		t.Fatalf("got unexpected error: %v", err.Error())
+	}
```

Comment:
The string value returned by the `Error()` method are _human readable_ messages, testing these message might create a high coupling between the test and the code being tested. For example, if we decide to change the prefix `interval:` to `intervalst`, because interval is too generic, and intervalst might reflect better the package name, we would have to change the test too. Depending on which unit test school you came from, that might not be a problem, but typically, [testing the error type is enough](https://quii.gitbook.io/learn-go-with-tests/questions-and-answers/error-types#problems-with-this-way-of-testing).

So I would either remove this test, or if you want to guarantee that at least the `from` and `to` values are part of this message, I would change this test to:
```suggestion
	if !strings.Contains(err.Error(), wantErr.from) {
		t.Fatalf("got error message: %q; want it to contain %q", err, wantErr.from)
	}
	
	if !strings.Contains(err.Error(), wantErr.to) {
		t.Fatalf("got error message: %q; want it to contain %q", err, wantErr.to)
	}
```

What do you think?

Reply:
I think I managed to do something simpler: [Error check refactoring](https://github.com/federicojasson/intervalst/commit/9b5123cab1b0dce21aae027f511c1ac6a7285584)

By using `errors.Is`, if I'm not mistaken, we can check both the type and the values (`from`/`to`) without having to hardcode the error string here.

Let me know if this makes sense.

Reply:
That seems great!